HPCC-17461 Rationalise the use of TimeElapsed

ghalliday opened the pull request:
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Minimal testing so far on a small sample of queries.

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

ghalliday committed:
HPCC-17461 Rationalise the use of TimeElapsed

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday commented:
@GordonSmith please take a look at this branch.  I think it should provide pairs for all graphs/subgraphs in hthor/roxie and thor.  It also switches compiler times to be total rather than elapsed.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 4cbff73d62a926487e3860e1f5280cf3072a951d
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.076 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |722 |722 |0 |
|test (thor) |617 |617 |0 |
|test (roxie) |746 |745 |1 |


Errors:
- test
  - roxie
    - schedule2, 

HPCC Stop: OK
HPCC Uninstall: OK

GordonSmith commented:
@ghalliday Will I ever see the a scope with all three of the following attributes:
* Started
* Finished
* TimeElapsed

Also do I consider it an error if I see a TimeElapsed without a Started?

Finally can I consider a Started + Finshed to be equal to a Started + TimeElapsed?

ghalliday commented:
Yes, there will be some scopes will all 3 defined.

In the long term there should always be a WhenStarted if there is a TimeElapsed.  One exception may be activities - which have a TimeFirstRow instead.  Short term you are likely to find quite a few which no corresponding WhenXXX

WhenStarted + TimeElapsed should equal WhenFinished.  However they are recorded at different degrees of accuracy: TimeElapsed is more accurate than WhenFinished-WhenStarted.

ghalliday commented:
See also https://github.com/hpcc-systems/LN/pull/838

richardkchapman commented:
Checklist is missing

ghalliday commented:
@richardkchapman now ready for review.  Rebased onto latest 6.4 and conflicts fixed.

richardkchapman commented on ecl/hthor/hthor.cpp:
```diff
@@ -236,8 +236,8 @@ void CHThorActivityBase::updateProgressForOther(IStatisticGatherer &progress, un
     progress.addStatistic(StNumRowsProcessed, numProcessed);
-    progress.addStatistic(StNumStarted, 1);  // wrong for an activity in a subquery
-    progress.addStatistic(StNumStopped, 1);
+    progress.addStatistic(StNumStarts, 1);  // wrong for an activity in a subquery
```
Is this comment still valid?

ghalliday replied:
Yes - it isn't keeping track of the number of times the child query has been started/stopped.

richardkchapman commented:
@ghalliday Needs to translate old names in existing workunits to new names. Should push back to 7.0.0 to get that done.

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 0b4e0d6afe1f7375bed370e03c528d20a1cc5cd8
Build: failed
Number of Error(s): 10
checking whether the -Werror option is usable... /mnt/disk1/home/vamosax/smoketest/PR-9898/HPCC-Platform/system/jlib/jstats.cpp:542:1: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-9898/HPCC-Platform/system/jlib/jstats.cpp:542:3: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-9898/HPCC-Platform/system/jlib/jstats.cpp:542:5: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-9898/HPCC-Platform/system/jlib/jstats.cpp:542:7: error: expected primary-expression before \xe2\x80\x98<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-9898/HPCC-Platform/system/jlib/jstats.cpp:542:9: error: \xe2\x80\x98HEAD\xe2\x80\x99 was not declared in this scope
/mnt/disk1/home/vamosax/smoketest/PR-9898/HPCC-Platform/system/jlib/jstats.cpp:543:5: error: expected \xe2\x80\x98}\xe2\x80\x99 before \xe2\x80\x98{\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-9898/HPCC-Platform/system/jlib/jstats.cpp:543:5: error: expected \xe2\x80\x98,\xe2\x80\x99 or \xe2\x80\x98;\xe2\x80\x99 before \xe2\x80\x98{\xe2\x80\x99 token
make[2]: *** [system/jlib/CMakeFiles/jlib.dir/jstats.cpp.o] Error 1
make[1]: *** [system/jlib/CMakeFiles/jlib.dir/all] Error 2
make: *** [all] Error 2
Get ECLWatch build result

ghalliday commented:
@GordonSmith do you specially look out for NumStarted on a graph?  Otherwise I can't think of any problems with renaming it - it isn't stored as text in a workunit.

@richardkchapman I have switched this back to 6.4 - this is a safer version.  Rather than reusing the previous enum I have added new values, which means the old names will be there.

ghalliday commented:
(Reasons for including in 6.4 is that it consistently adds start/end times to graphs/subgraphs.  But I am happy to rebase to master if preferred or still too much potential for problems.)  Main problem I can think of is TimeElapsed replaced with TimeLocalExecute for timings from eclccserver.

GordonSmith commented:
@ghalliday I do special case "NumStarted" and "NumStopped" in js and c++ GraphControl

ghalliday commented:
ok.  I'll retarget to master, and open a couple of jiras to ensure the graph control spots the new ones

ghalliday commented:
@richardkchapman rebased, and a jira updated/isue opened on graph control.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 8dd9472a3efe6099c3f5a01f24b45cdbd818871b
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.121 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |727 |727 |0 |
|test (thor) |620 |620 |0 |
|test (roxie) |748 |747 |1 |


Errors:
- test
  - roxie
    - schedule2, 

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 66c0b6852482185249a3b224f095f1c1c6654792
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.047 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |626 |626 |0 |
|test (roxie) |755 |754 |1 |


Errors:
- test
  - roxie
    - schedule2, 

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 66c0b6852482185249a3b224f095f1c1c6654792
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.675 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |736 |736 |0 |
|test (thor) |630 |630 |0 |
|test (roxie) |765 |764 |1 |


Errors:
- test
  - roxie
    - schedule2, 

HPCC Stop: OK
HPCC Uninstall: OK
